### PR TITLE
Duplicate VS2013 logic for tie breaking overloads based on param-array types when none of the arguments matched the parameter.

### LINKED ISSUE #4458: Code that compiles in VS2013 fails with CS0121 in 2015; overloads with different params parameter types
The following code compiles in VS2013 but in VS2015 it produces the following compiler error:

> CS0121: The call is ambiguous between the following methods or properties: 'Test.Test(int, params string[])' and 'Test.Test(int, params Expression<string>[])'

```cs
    using System.Linq.Expressions;

    namespace ConsoleApplication9
    {
        internal class Program
        {
            private static void Main(string[] args)
            {
                var x = new Test(10);
            }
        }

        public class Test
        {
            public Test(int a, params string[] p) { }
            public Test(int a, params Expression<string>[] p) { }
        }
    }
```

In VS2013 the call will be routed to the expression overload (observational fact, not 100% sure that will *always* happen though)

Arguably, this is not good design but it is still a backwards compatibility issue.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #4458.

@gafter, @VSadov, @jaredpar, @agocke Please review.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs [hunk 1]
```diff
@@ -1543,6 +1562,90 @@ private static TypeSymbol GetParameterType(int argIndex, MemberAnalysisResult re
             return BetterResult.Neither;
         }
 
+        private static BetterResult BetterParamArrayTypeTieBreaker(TypeSymbol type1, TypeSymbol type2)
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs]: This doesn't seem very rational. I would like us to consider whether we can break compat here for more rational overload resolution.

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs [hunk 2]
```diff
@@ -1543,6 +1562,90 @@ private static TypeSymbol GetParameterType(int argIndex, MemberAnalysisResult re
...
+                            return BetterResult.Left;
+                        }
+                        else if (builder1.Count < builder2.Count)
+                        {
+                            return BetterResult.Right;
+                        }
+
+                        BetterResult result = BetterResult.Neither;
+
+                        for (int i = 0; i < builder1.Count; i++)
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs]: I understand what this is doing but I can't think of any rational explanation for why it should be this way.

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs]: @MadsTorgersen Can you convince yourself that this behavior is worth retaining?

**Comment by diryboy** [comment]
This kind of rule is confusing to me.

In VS2013, consider
```C#
static void Main( string[] args )
{
    Some();
}
```

If I can understand that
`static void Some( params Action<Func<int, int>, int>[] funcs )`
 is more specific than
`static void Some( params Action<Func<int>, int>[] funcs )`

and that
`static void Some( params Action<Func<int, int>, int>[] funcs )`
 is ambiguous with
`static void Some( params Action<int, Func<int, int>>[] funcs )`

Then I cannot understand why
`static void Some( params Action<Func<int>, int>[] funcs )`
 is ambiguous with
`static void Some( params Action<int, Func<int, int>>[] funcs )`

**Comment by AlekseyTs** [comment]
In the interest of the future health of our overload resolution code, we dicided to keep the breaking (and correct) behavior. If we get more than this single case, we may want to reevaluate.

**Pull request closed without merging**